src/Microsoft.Health.Dicom.SqlServer/Features/Workitem/WorkitemSqlQueryGenerator.cs

```diff
@@ -0,0 +1,179 @@
+using Microsoft.Health.Dicom.SqlServer.Features.Query;
+using Microsoft.Health.Dicom.SqlServer.Features.Schema;
+using Microsoft.Health.Dicom.SqlServer.Features.Schema.Model;
+using Microsoft.Health.SqlServer;
+using Microsoft.Health.SqlServer.Features.Schema.Model;
+using Microsoft.Health.SqlServer.Features.Storage;
+
+namespace Microsoft.Health.Dicom.SqlServer.Features.Workitem
+{
+    internal class WorkitemSqlQueryGenerator : BaseSqlQueryGenerator
+    {
+        private readonly IndentedStringBuilder _stringBuilder;
+        private readonly QueryExpression _queryExpression;
+        private readonly SchemaVersion _schemaVersion;
+        private const string WorkitemTableAlias = "w";
+
+        public WorkitemSqlQueryGenerator(
+            IndentedStringBuilder stringBuilder,
+            QueryExpression queryExpression,
+            SqlQueryParameterManager sqlQueryParameterManager,
+            SchemaVersion schemaVersion,
+            int partitionKey)
+            : base(stringBuilder, sqlQueryParameterManager, partitionKey)
+        {
+            _stringBuilder = stringBuilder;
+            _queryExpression = queryExpression;
+            _schemaVersion = schemaVersion;
+
+            if ((int)_schemaVersion < SchemaVersionConstants.SupportUpsRsSchemaVersion)
+            {
+                throw new BadRequestException(DicomSqlServerResource.SchemaVersionNeedsToBeUpgraded);
+            }
+
+            Build();
+        }
+
+        private void Build()
+        {
+            string projectionTableAlias = "f";
+
+            AppendSelect(projectionTableAlias);
+
+            AppendFilterTable(projectionTableAlias);
+
+            AppendOptionRecompile();
+        }
+
+        private void AppendFilterTable(string filterAlias)
+        {
+            _stringBuilder
+                .AppendLine("( SELECT ")
+                .Append(VLatest.Workitem.WorkitemUid, WorkitemTableAlias).AppendLine(",")
+                .Append(VLatest.Workitem.WorkitemKey, WorkitemTableAlias).AppendLine()
+                .AppendLine($"FROM {VLatest.Workitem.TableName} {WorkitemTableAlias}");
+
+            AppendExtendedQueryTagTables();
+
+            _stringBuilder.AppendLine("WHERE 1 = 1");
+
+            AppendPartitionWhereClause(WorkitemTableAlias);
+
+            using (IndentedStringBuilder.DelimitedScope delimited = _stringBuilder.BeginDelimitedWhereClause())
+            {
+                AppendFilterClause();
+            }
+
+            AppendFilterPaging();
+
+            _stringBuilder.AppendLine($") {filterAlias}");
+        }
+
+        private void AppendPartitionWhereClause(string tableAlias)
+        {
+            if ((int)_schemaVersion >= SchemaVersionConstants.SupportDataPartitionSchemaVersion)
+            {
+                _stringBuilder.AppendLine($"AND {tableAlias}.{VLatest.Partition.PartitionKey} = {PartitionKey}");
+            }
+        }
+
+        private void AppendExtendedQueryTagTables()
```

Comment:
Should this be extracted to a common class? Seems the logic is same for generating query on these tables maybe the keys are different.

Reply:
Yeah extracted.